target/linux/ath79/dts/qca9558_librerouter-v1.dts

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca9557.dtsi"
+
+/ {
+	compatible = "librerouter,librerouter-v1", "qca,qca9557";
+	model = "LibreRouter v1";
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
+
+	aliases {
+		led-boot = &system;
+		led-failsafe = &system;
+		led-running = &system;
+		led-upgrade = &system;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		system: system {
+			label = "librerouter:green:system";
+			gpios = <&gpio 23 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "heartbeat";
+		};
+
+		wifi_green {
+			label = "librerouter:green:wlan2g";
+			gpios = <&gpio 22 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "Reset button";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		gpio_poe_passthrough {
+			gpio-export,name = "librerouter:power:poe_pass";
+			gpio-export,output = <0>;
+			gpios = <&gpio 1 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	watchdog {
+		compatible = "linux,wdt-gpio";
+		gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		hw_algo = "toggle";
+		hw_margin_ms = <1000>;
+		always-running;
+	};
+
+};
+
+&pcie0 {
+	status = "okay";
+};
+
+&pcie1 {
+	status = "okay";
+};
+
+&uart {
+	status = "okay";
+};
+
+&gpio {
+	status = "okay";
+};
+
+&usb_phy0 {
+	status = "okay";
+};
+
+&usb0 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	status = "okay";
+
+	hub_port0: port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
+};
+
+&usb_phy1 {
+	status = "okay";
+};
+
+&usb1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	status = "okay";
+
+	hub_port1: port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
+};
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x040000>;
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
```

Comment:
Since you are not using the uboot-envtools make this partition read-only.

Reply:
Yes the u-boot-env partition is used for dual-boot support (not included in this PR), it should be stated somewhere?

Reply:
Do you think you please could include it?

Reply:
Dual boot support for this device needs `CONFIG_MIPS_CMDLINE_DTB_EXTEND=y` or `CONFIG_MIPS_CMDLINE_FROM_BOOTLOADER=y`. Do you know if there is a way to enable this options only for this device? Beacause  AFAIK the other ath79 devices may break changing this config.

Reply:
> Dual boot support for this device needs CONFIG_MIPS_CMDLINE_DTB_EXTEND=y or CONFIG_MIPS_CMDLINE_FROM_BOOTLOADER=y. Do you know if there is a way to enable this options only for this device? Beacause AFAIK the other ath79 devices may break changing this config.

No, not with the MIPS target. However, it's possible to do that with PowerPC (yeah, this is a bit silly).

Now, Ideally you would want to support FIT so you could pass the dtb to the bootloader to make the changes to the dtb. However, you bootloader source is too old for this trick. However, it seems to support the multi-image, which can also be used to pass a dtb to the bootloader to be edited. What I don't know is if the ath79 MIPS linux kernel init code can deal with having the bootloader passing on the dtb.

Reply:
I will investigate FIT and multi-image, thanks! Dual boot support is not mandatory. The device is perfectly usable without using the dual-boot support and the images generated work fine. 

We may maintain out of tree support for dual-booting as also it is implemented using a the userspace tool `safe-upgrade` (https://github.com/libremesh/lime-packages/blob/master/packages/safe-upgrade/files/usr/sbin/safe-upgrade) that is not included in openwrt. dual-booting feature can be installed/bootstraped by the user using `safe-upgrade` so from factory the openwrt images will work. If the device has the dual-boot installed then installing images without cmdline support from bootloader won't work but the device won't brick (as it has automatic rollback).

Reply:
I am marking this as resolved because I don't think we can do more at least at this moment. Please reopen if you think otherwise